File: echosounder/src/depth_publisher.cpp
/*------------------------------------------------------------------------------
* ROS node "depth_publisher" connects to the Tritech Micron Echocounder and
* publishes the depth of the seabed on the topic "depth".
* ------------------------------------------------------------------------------
*/

#include <ros/ros.h>
#include <string.h>
#include "serial/serial.h"
#include <iostream>
#include "std_msgs/Float64.h"

using namespace std;

serial::Serial ser;

ros::Publisher depth_pub;

int main(int argc, char **argv)
{
    ros::init(argc, argv, "depth_publisher");
    ros::NodeHandle nh;
    ros::NodeHandle nhp("~");

    depth_pub = nh.advertise<std_msgs::Float64>("depth", 1000);

    string port;
    nhp.param<string>("port", port, "/dev/ttyUSB0");

    try
    {
	ser.setPort(port);
	ser.setBaudrate(9600);
	serial::Timeout to = serial::Timeout::simpleTimeout(1000);
	ser.setTimeout(to);
	ser.open();
    }
    catch (serial::IOException& e)
    {
	ROS_ERROR_STREAM("Unable to open sonar port");
	return -1;
    }

    if (ser.isOpen())
    {
	ROS_INFO_STREAM("Serial port of the sonar initialized");
    }
    else return -1;


    ros::Rate loop_rate(2);

    while (ros::ok())
    {	
//        depth_pub.publish(depth_msg);
        ros::spinOnce();

        if(ser.available()){
            ROS_DEBUG_STREAM("Reading from serial port");
	    string msg = ser.read(ser.avaible());
	    string depth = msg.substr(6,9);
            std_msgs::Float64 depth_msg;
            depth_msg.data = depth;
	    ROS_DEBUG_STREAM("Read: " << depth_msg.data);
            depth_pub.publish(depth_msg);
        }
        loop_rate.sleep();

    }
    return 0;
}

File: echosounder/src/depth_publisher.py
#!/usr/bin/env python

import rospy
import serial
from std_msgs.msg import Float64

rospy.init_node("depth_publisher")
depth_pub = rospy.Publisher("depth", Float64, queue_size=1)

ser = serial.Serial(port="/dev/sonar", baudrate=9600)

rate = rospy.Rate(5)
if ser.isOpen():
    while not rospy.is_shutdown():
        line = ser.readline()
        if line[0] == '0':
            depth_msg = Float64()
            depth_msg.data = float(line.split('m')[0])
            depth_pub.publish(depth_msg)
        rate.sleep()
    ser.close()

else:
    rospy.logwarn("Error opening sonar port")
    exit()
